### PR TITLE
RE-7 Add documentation for debugging

### PR DESCRIPTION
**Pull request by hughsaunders**
This documentation is intended to help people who are adding jobs or
functionality to existing to debug problems.

Issue: [RE-7](https://rpc-openstack.atlassian.net/browse/RE-7)

### CODE REVIEW

File: docs/debugging.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# Debugging Gate Failures
+
+## Blue Ocean
+[Blue ocean](https://rpc.jenkins.cit.rackspace.net/blue/organizations/jenkins/pipelines) is the new UI for [Jenkins](https://rpc.jenkins.cit.rackspace.net/), it makes it easier to find the point where
+a job failed, without trawling through the console. To use blue ocean, follow
+the link from a github pull request status to a job, then click open "Open Blue
```
Comment (mancdaz) [inline docs/debugging.md]: redundant use of open

File: docs/debugging.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
+# Debugging Gate Failures
+
+## Blue Ocean
+[Blue ocean](https://rpc.jenkins.cit.rackspace.net/blue/organizations/jenkins/pipelines) is the new UI for [Jenkins](https://rpc.jenkins.cit.rackspace.net/), it makes it easier to find the point where
+a job failed, without trawling through the console. To use blue ocean, follow
```
Comment (mancdaz) [inline docs/debugging.md]: ...where a job failed without trawling.... (remove comma)

File: docs/debugging.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
+# Debugging Gate Failures
+
+## Blue Ocean
+[Blue ocean](https://rpc.jenkins.cit.rackspace.net/blue/organizations/jenkins/pipelines) is the new UI for [Jenkins](https://rpc.jenkins.cit.rackspace.net/), it makes it easier to find the point where
```
Comment (mancdaz) [inline docs/debugging.md]: ...*that* makes it easier...

File: docs/debugging.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+a job failed, without trawling through the console. To use blue ocean, follow
+the link from a github pull request status to a job, then click open "Open Blue
+Ocean" from the left sidebar.
+
+<img src="images/debugging/sidebar.png" width=200>
+
+Blue ocean displays a horizontal pipeline of stages. Stages that failed will
+be shown in red. Clicking a stage reveals a list of steps beneath,
```
Comment (mancdaz) [inline docs/debugging.md]: repeated statement of fact that 'stages that fail will be shown red'.

File: docs/debugging.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+steps that failed will also be marked as red. Expanding a step will show it's logs.
```
Comment (mancdaz) [inline docs/debugging.md]: s/it's/its/

File: docs/debugging.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+![](images/debugging/blueocean.png)
+
+Using this method is much quicker for finding the failure point than reading the
+console.
+
+## The Console
+Jenkins collects all the output of commands it executes, these outputs are
+streamed to the console for each Job. This includes logging for Jenkins
```
Comment (mancdaz) [inline docs/debugging.md]: ...*Jenkins collects all of the output from commands it executes, and these outputs*...

File: docs/debugging.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+operations such as creating a single use instance for a project's jobs to
+execute on. As the console includes logging for steps not defined by a project,
+its inefficient to wade through it when looking for the cause of a failure.
```
Comment (mancdaz) [inline docs/debugging.md]: s/its/it is/

File: docs/debugging.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+## Artefacts
+While command outputs as viewed in blue ocean or the console are useful, they
+don't tell the whole story. Jobs can produce artefacts which are stored
+for a [set time](https://github.com/rcbops/rpc-gating/blob/master/playbooks/upload_to_cloud_files.yml#L13-L19) after the completion of a job. For RPC-O AIO builds, these
+artefacts include the etc and log dirs of each container and the host.
+
```
Comment (mancdaz) [inline docs/debugging.md]: Given that at some point the artefacts that get stored for a particular job type should be definable by the team that is responsible for the software, maybe we should genericise this section and add a separate note to indicate RPCO specifics (which we will remove once we get to the point that teams define their own artefacts)

File: docs/debugging.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+To view artefacts, click "Build Artifact Links" in the sidebar from the standard
+Jenkins UI page for a build. This leads to an artefact index page which usually
+contains a single file; a tar of all the available logs and config.
+
```
Comment (mancdaz) [inline docs/debugging.md]: ...*a single file, which is a tar*...

File: docs/debugging.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+to login to a gate instance and investigate the problem. Normally instances
+are deleted as soon as they finish executing a job. In order to retain
+the instance, the job must be rerun with "Pause" added to the STAGES parameter.
+![](images/debugging/add_pause.png)
+Note that it's not possible to change parameters via a pull request comment
+recheck command, so the rerun will have to be triggered from the Jenkins UI. To
+trigger a build select 'Rebuild' (not replay!) from the side bar.
+
+Adding Pause will cause Jenkins to wait for input before clearing up the node.
+This will not retain the node indefinitely, it will be cleaned up by a periodic
```
Comment (mancdaz) [inline docs/debugging.md]: ...*indefinitely, because it will be*...

File: docs/debugging.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+job when it hits the instance [age
+limit](https://github.com/rcbops/rpc-gating/blob/master/rpc_jobs/periodic_cleanup.yml#L6-L9).
+It's important to continue the paused Jenkins job after you have finished
+investigating, as even if the instance is deleted the Job will be reserving at
+least one Jenkins executor. To resume a job, view the console for the running job,
+the last line will contain links for continue and abort. If you don't have
+permissions to resume the job ping us in [slack #rpc-gating ](https://rackspace.slack.com/messages/C0G5K94RJ)
+
```
Comment (mancdaz) [inline docs/debugging.md]: #rpc-releng should probably become the equivalent of #openstack-infra, and it's where the team will be present to support everything RE related, not just gating.

File: docs/debugging.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+```
+Stage Complete: Connect Slave
+[Pipeline] }
+[Pipeline] // stage
+[Pipeline] node
+Running on raltdcp-10-77a6 in /var/lib/jenkins/workspace/RPC-AIO_liberty-trusty-deploy-ceph-pr
+```
+
+Note the node name after "running on", view the config for that node:
```
Comment (mancdaz) [inline docs/debugging.md]: *Note the node name after "running on". To view the config for that node, replace YOUR-NODE-NAME with the node name you noted in the last step*

File: docs/debugging.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+next section.
+
+### SSH keys
+All the public keys in the [keys file](https://github.com/rcbops/rpc-gating/blob/master/keys/rcb.keys) are injected
+into root's authorized keys for all instances used for testing. If you need
+access to these instances, create a PR against that file to add your key.
+
+
+## Contact Us
+If you have any issues, please ask the team for help in [slack #rpc-gating](https://rackspace.slack.com/messages/C0G5K94RJ)
```
Comment (mancdaz) [inline docs/debugging.md]: rpc-releng

File: docs/debugging.md (hunk 4 above)
Comment (hughsaunders) [inline docs/debugging.md]: Steps vs stages

File: docs/debugging.md (hunk 3 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 2 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 1 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 6 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 7 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 8 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 9 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 10 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 11 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 12 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 13 above)
Comment (hughsaunders) [inline docs/debugging.md]: Done

File: docs/debugging.md (hunk 4 above)
Comment (hughsaunders) [inline docs/debugging.md]: Decided to reword this paragraph, Done.

File: docs/debugging.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+don't tell the whole story. Jobs can produce artifacts which are stored
+for a [set time](https://github.com/rcbops/rpc-gating/blob/master/playbooks/upload_to_cloud_files.yml#L13-L19) after the completion of a job. Any job can produce artifacts
+and use the provided functions to publish those. Jobs artifacts should include
+everything required to debug a build.
+
+To view artifacts, click "Build Artifact Links" in the sidebar from the standard
+Jenkins build page.
+
+### Artefact Example: RPCO
+
```
Comment (mancdaz) [inline docs/debugging.md]: Artifact, for consistency. Sorry, missed this first time.

**Pull request merged**